File: novi-app/activations/src/main/java/org/novi/activations/WeightedRandomActivation.java
package org.novi.activations;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.commons.math3.distribution.EnumeratedDistribution;
import org.apache.commons.math3.random.JDKRandomGenerator;
import org.apache.commons.math3.random.RandomGenerator;
import org.apache.commons.math3.util.Pair;
import org.novi.core.activations.BaseActivation;
import org.novi.core.exceptions.ConfigurationParseException;

import java.util.List;
import java.util.Map;

public class WeightedRandomActivation implements BaseActivation<List<Pair<String, Double>>> {

    private List<Pair<String, Double>> configuration;

    @Override
    public WeightedRandomActivation configuration(String configuration) throws ConfigurationParseException {
        try {
            TypeReference<Map<String, Double>> tref = new TypeReference<>() {
            };
            ObjectMapper mapper = new ObjectMapper();
            Map<String, Double> parsedConfig = mapper.readValue(configuration, tref);
            this.configuration = parsedConfig.entrySet().stream().map(e -> Pair.create(e.getKey(), e.getValue())).toList();
            return this;
        } catch (JsonProcessingException e) {
            throw new ConfigurationParseException(e);
        }
    }

    @Override
    public Boolean apply(String context){
        try {
            ObjectMapper mapper = new ObjectMapper();
            JsonNode root = mapper.readTree(context);
            Map<String, Object> contextMap = mapper.treeToValue(root, Map.class);
            return evaluateFor(contextMap);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    public Boolean evaluateFor(Map<String, Object> context) {
        Map<String, Object> contextMap = (Map<String, Object>) context.get(getName());
        int seed = (int) contextMap.get("seed");
        String variantToCheck = (String) contextMap.get("variantToCheck");
        RandomGenerator rnd = new JDKRandomGenerator(seed);
        EnumeratedDistribution<String> ed = new EnumeratedDistribution<>(rnd, this.configuration());
        return ed.sample().equalsIgnoreCase(variantToCheck);
    }

    @Override
    public List<Pair<String, Double>> configuration() {
        return configuration;
    }
}

File: novi-activations/src/main/scala/org/novi/activations/dsl/WeightedRandomActivation.scala
package org.novi.activations.dsl

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.core.`type`.TypeReference
import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
import org.apache.commons.math3.distribution.EnumeratedDistribution
import org.apache.commons.math3.random.JDKRandomGenerator
import org.apache.commons.math3.util.Pair
import org.novi.core.activations.{BaseActivation, BaseActivationFactory}
import org.novi.core.dsl.DslActivation

import java.util.{Collections, List, Map}
import java.{lang, util}

class WeightedRandomActivation(configuration: util.List[Pair[String, lang.Double]]) extends DslActivation(configuration) {
  def this() = {
    this(null)
  }

  override def evaluate(context: String): lang.Boolean = try {
    val mapper: ObjectMapper = new ObjectMapper
    val root: JsonNode = mapper.readTree(context)
    val parsedContext: util.Map[String, AnyRef] = mapper.treeToValue(root, classOf[util.Map[String, AnyRef]])
    val contextMap = parsedContext.get(getName).asInstanceOf[util.Map[String, AnyRef]]
    val seed = contextMap.get("seed").asInstanceOf[Int]
    val variantToCheck = contextMap.get("variantToCheck").asInstanceOf[String]
    val rnd = new JDKRandomGenerator(seed)
    val ed = new EnumeratedDistribution[String](rnd, this.configuration)
    ed.sample.equalsIgnoreCase(variantToCheck)
  } catch {
    case e: JsonProcessingException =>
      throw new RuntimeException(e)
  }


  override def apply(configuration: String): BaseActivation[util.List[Pair[String, lang.Double]]] = WeightedRandomActivation.apply(configuration)
}

object WeightedRandomActivation extends BaseActivationFactory[util.List[Pair[String, lang.Double]]]{
  override def apply(configuration: String): WeightedRandomActivation = {
    val variant_and_weights: util.List[Pair[String, lang.Double]] = if (configuration != null) {
      val tref: TypeReference[util.Map[String, lang.Double]] = new TypeReference[util.Map[String, lang.Double]]() {}
      val mapper = new ObjectMapper
      val parsedConfig: util.Map[String, lang.Double] = mapper.readValue(configuration, tref)
      parsedConfig.entrySet.stream.map((e: util.Map.Entry[String, lang.Double]) => Pair.create(e.getKey, e.getValue)).toList
    } else {
      Collections.emptyList()
    }
    new WeightedRandomActivation(variant_and_weights)
  }

}